addons/purchase_mrp/models/mrp_bom.py

```diff
@@ -21,6 +21,31 @@ def _check_bom_lines(self):
                 raise UserError(_("The total cost share for a BoM's component have to be 100"))
         return res
 
+    def _prepare_bom_done_values(self, quantity, product, original_quantity, current_line, boms_done):
+        result = super()._prepare_bom_done_values(quantity, product, original_quantity, current_line, boms_done)
+        result['bom_cost_share'] = self._get_line_cost_share(boms_done, current_line)
+        return result
+
+    def _prepare_line_done_values(self, quantity, product, original_quantity, parent_line, current_line, boms_done):
+        result = super()._prepare_line_done_values(quantity, product, original_quantity, parent_line, current_line, boms_done)
+        result['line_cost_share'] = float_round(self._get_line_cost_share(boms_done, current_line), precision_digits=2)
+        return result
+
+    def _get_line_cost_share(self, boms_done, current_line):
+        if not current_line:
+            return 100.0
+        parent_boms_done = next(((bom_entry, values) for bom_entry, values in reversed(boms_done) if bom_entry == current_line.bom_id), None)
+        parent_cost_share = 100.0 if not parent_boms_done else parent_boms_done[1].get('bom_cost_share', 100.0)
+        line_cost_share = parent_cost_share * (current_line._get_cost_share())
+        return line_cost_share
+
+    def _round_last_line_done(self, lines_done):
+        result = super()._round_last_line_done(lines_done)
+        if result:
+            last_vals = next((vals for _, vals in reversed(result) if not float_is_zero(vals.get('line_cost_share', 0), precision_digits=2)), result[-1][1])
+            last_vals['line_cost_share'] = float_round(100.0 - (sum(vals.get('line_cost_share', 0) for _, vals in lines_done) - last_vals.get('line_cost_share', 0)), precision_digits=2)
```

Comment:
I think we should keep it simple. Also you should use the result rather than lines_done to avoid ignoring the  super call (imagine an other module needs to override this method some day and already provided a rounded result for the `line_cost_share` on an other line, you should not forget it and add a new rounding with your computation based on the lines_done state). What about this instead ?
```suggestion
            result[-1][1]['line_cost_share'] = float_round(100.0 - sum(vals.get('line_cost_share', 0) for _, vals in result[:-1]), precision_digits=2)
```